### ISSUE TITLE
Let Flow remove illegal (forbidden) characters (\ / : * ? " ' < > |) in the query string

### ISSUE THREAD

**Issue by raindropsfromsky** [Flow-Launcher/Flow.Launcher#1771]
A user often copies the title of an article/book from internet, and then pastes it in the Flow to check if that file exists.

However, the problem with that approach is that often the text string contains non-Windows characters such as a column (:).
Earlier, Flow was able to ignore such characters. 
But now looks for a positive match.
As a result, Flow hides even when there are matching results, just because they cannot have the : character.
If the user deletes such "non-Windows" characters, the results show up.

Therefore, it would be great if Flow ignores such characters.

**Comment by taooceros** [comment]
I am confused about what do you mean by "Non-Windows characters". I don't think a `:` will prevent results showing (maybe will prevent results from explorer).

**Comment by z1nc0r3** [comment]
You probably mean the **forbidden characters** for file/dir names like `\ / : * ? " ' < > |`. And yes, FL considers these characters for searching files/dirs and will not give results correctly when the query contains these illegal characters. 

I think we can just replace these forbidden chars with "" in the explorer plugin. So it'll not consider these chars for searching files/dirs.

**Comment by taooceros** [comment]
> You probably mean the **forbidden characters** for file/dir names like `\ / : * ? " ' < > |`. And yes, FL considers these characters for searching files/dirs and will not give results correctly when the query contains these illegal characters.
> 
> I think we can just replace these forbidden chars with "" in the explorer plugin. So it'll not consider these chars for searching files/dirs.

yeah that should be the case. I will check recently

**Comment by raindropsfromsky** [comment]
> You probably mean the **forbidden characters** for file/dir names like `\ / : * ? " ' < > |`. And yes, FL considers these characters for searching files/dirs and will not give results correctly when the query contains these illegal characters.
> 
> I think we can just replace these forbidden chars with "" in the explorer plugin. So it'll not consider these chars for searching files/dirs.

Yes, that's what I meant. Thanks for the correction! 
I have corrected the subject line of my original issue.

Yes, FL should simply delete (or ignore) these characters and search for the rest.

**Comment by raindropsfromsky** [comment]
BTW even **Everything** has the same issue: It cannot ignore illegal (forbidden) characters in the search string. 

So regardless of which engine is used (Explorer or Everything), the result would be the same: There is a false negative. That is, although there are matching results, they are not reported.

Therefore the best strategy is for FL to pre-process the string and remove all forbidden characters.
If FL passes pre-corrected search string, all indexing engine(s) will be able to find all results without any issue.

**Comment by raindropsfromsky** [comment]
Wow! Is this already addressed?
How soon can check out the modified alpha/beta/nightly?
Please share the download link!

**Comment by z1nc0r3** [comment]
@raindropsfromsky Since this issue needs a few more changes I closed the PR.

**Comment by z1nc0r3** [comment]
@raindropsfromsky Can you please try using the **Windows Index** as the search engine? Surprisingly I found that the **Windows search engine** handles forbidden characters already. #1795 I added a filter for the **Everything** engine.

**Comment by raindropsfromsky** [comment]
Yes, I do remember that FL used to ignore the presence of such characters.
Probably my old installation (in Win 10) made use of Windows Indexing only.

But after upgrade to Win 11, I am unable to get that result.

I tried your advise, and here's what happened:
I set **Windows Index** as default (see the screenshot below).
![image](https://user-images.githubusercontent.com/9047168/212469363-3b51dd74-38d4-4926-829a-c04e7b3e836c.png)

But I am not getting ANY results whatsoever now.
I had faced this issue earlier (#1767) after upgrade to Win 11, but it had somehow sorted out after I changed the number of results to 15 from the default 5.

But this time, I tried the same trick, but the problem does not go away.

So I changed all three options to **Everything**, but now there is a new problem: FL is unable to integrate with the **Everything** service.

I am getting a warning that Everything service is not running.
![image](https://user-images.githubusercontent.com/9047168/212469713-3430bb3e-aba1-48e5-b811-f7c0d301768f.png)

In the past, I used to click on the warning to start the service, but this time, even when I click on the warning, the **Everything** service is launched, but FL keeps showing the same warning.

I can launch Everything windows from the System Tray and use it independently to search for any file/folder.
But FL fails to sense that the **Everything** service has started.

**Comment by raindropsfromsky** [comment]
I uninstalled both Everything and FL, and then re-installed FL.
But the same problem repeats.

**Comment by VictoriousRaptor** [comment]
@raindropsfromsky Are you using Everything v1.5a? Currently Flow doesn't support it.

**Comment by raindropsfromsky** [comment]
> @raindropsfromsky Are you using Everything v1.5a? Currently Flow doesn't support it. You need to modifiy some config file to make it compatible.

Well, I expected FL to start **Everything** as a service (it used to do that previously).
In the past, I clicked on the warning, and a dialog popped up, in which I selected "no", to let FL install the Everything service automatically.

But since this does not work any more, I downloaded Everything from its website and installed it.
The Version is 1.4.1.1009 (x64).

But even that stand-alone installation is not getting registered in FL.

**Comment by z1nc0r3** [comment]
> But I am not getting ANY results whatsoever now.
> I had faced this issue earlier (https://github.com/Flow-Launcher/Flow.Launcher/issues/1767) after upgrade to Win 11, but it had somehow sorted out after I changed the number of results to 15 from the default 5.

For this, please check your windows index settings and confirm that your files are indexed. You can also try `Rebuild` option on indexing settings (takes some time).

> But this does not work any more, so I downloaded Everything from its website and installed it.
> The Version is 1.4.1.1009 (x64).

For this please try using this setting in **Everything application**. Hope you already have uninstalled the **Everything plugin** from the _FL > Plugins_

![image](https://user-images.githubusercontent.com/64279853/212471176-32455852-1e5f-48ff-b3f4-42c8d14d9f1b.png)

**Comment by raindropsfromsky** [comment]
Thanks for the tips.

I rebuilt the Windows index, and now I am able to see results. 

BTW my **Windows index** dialog shows only 2823 items. I wonder if that's a reasonable number, given that I have connected five HDDs with >15 TB of data (mostly pdf, odt, epub and mp4 files).

In **Everything** options, the **Everything service** checkbox was already selected.

But when I tried to uninstall the **Everything** plugin, but I could not find it in the **Plugin** page or on the **Plugin store** page. So I tried to enter the **pm uninstall** command in the FL's search bar. But it does not show the **Everything** plugin at all.

**Comment by z1nc0r3** [comment]
> I rebuilt the Windows index, and now I am able to see results. But only when I select Everything as the engine in the Explorer plugin. If I switch over to the Windows index engine, the results disappear.

Windows Indexing usually takes a long time to complete (depending on the storage type, and capacity). Since you 
have HDD with over 15TB, indexing should probably take lots of time. So I'm not sure, whether the indexing is complete. And also the file count should be more than 2823.
_One thing, Windows Indexing does not affect the **Everything** engine. **Everything** has its indexing function. You can see the indexed file count in the **Everything** application's bottom left corner._

> But when I tried to uninstall the Everything plugin, but I could not find it in the Plugin page or on the Plugin store page. So I tried to enter the pm uninstall command in the FL's search bar. But it does not show the Everything plugin at all.

Then you should have already uninstalled the Everything plugin and no need to check again. Please try reinstalling the **Everything** application again.

**Comment by raindropsfromsky** [comment]
@z1nc0r3 

**Update:** 

Based on your advise, I checked the **Indexing options** dialog in my newly installed Windows 11.
It only listed a single node named **Start Menu**. 

I clicked on the **Modify** button.
This allowed me to add all the HDDs (including external HDDs).
Then I restarted indexing.

For the last two days, the indexing is going on, but it seemed very slow.
In fact, the dialog always reported "Indexing is paused" status.

I googled for a reason, and found this [article](https://office-watch.com/2022/force-windows-indexing-to-run-faster).
It says that indexing pauses whenever we work on the laptop.
To speed up indexing, we have to let it run in the background.

I followed this process, and then restarted the **Windows search** service.
Now the indexing is running continuously even when I am working on the laptop.

I hope to get all my HDDs indexed soon, which will enable me to use the **Windows Index** (instead of **Everything**).
You have already pointed out that  Windows Index can handle the forbidden (illegal) characters.

Once the indexing is finished, I will check this out and confirm it here.

Thanks for all the help!

**Comment by raindropsfromsky** [comment]
Update:

The indexing is still going on, but I tested and confirmed that **windows explorer** can ignore the illegal characters.

That said, this feature is NOT available if we are using Everything. 
Therefore, this feature should be built in FL itself.

**Comment by taooceros** [comment]
hmm do you experiment this error?

```

Please open new issue in: https://github.com/Flow-Launcher/Flow.Launcher/issues/new
1. upload log file: C:\Users\1\AppData\Roaming\FlowLauncher\Logs\1.11.0\2023-01-22.txt
2. copy below exception message

Flow Launcher version: 1.11.0
OS Version: Microsoft Windows NT 10.0.22623.0
IntPtr Length: 8
x64: True

Python Path: C:\Users\1\AppData\Roaming\FlowLauncher\Environments\Python\PythonEmbeddable-v3.8.9\pythonw.exe
Node Path: 
Date: 01/22/2023 20:16:29
Exception:
Explorer Exception: 
Websites: https://github.com/Flow-Launcher/Flow.Launcher
Author: Jeremy Wu
Version: 2.1.0
Flow.Launcher.Core.ExternalPlugins.FlowPluginException: Object reference not set to an instance of an object.
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Flow.Launcher.Plugin.Explorer.Search.SearchManager.<>c__DisplayClass4_0.<SearchAsync>b__1(AccessLink excludedPath) in C:\projects\flow-launcher\Plugins\Flow.Launcher.Plugin.Explorer\Search\SearchManager.cs:line 120
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at System.Collections.Generic.HashSet`1.RemoveWhere(Predicate`1 match)
   at Flow.Launcher.Plugin.Explorer.Search.SearchManager.SearchAsync(Query query, CancellationToken token) in C:\projects\flow-launcher\Plugins\Flow.Launcher.Plugin.Explorer\Search\SearchManager.cs:line 119
   at Flow.Launcher.Plugin.Explorer.Main.QueryAsync(Query query, CancellationToken token) in C:\projects\flow-launcher\Plugins\Flow.Launcher.Plugin.Explorer\Main.cs:line 88
   at Flow.Launcher.Core.Plugin.PluginManager.<>c__DisplayClass22_0.<<QueryForPluginAsync>b__0>d.MoveNext() in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 191
--- End of stack trace from previous location ---
   at Flow.Launcher.Infrastructure.Stopwatch.DebugAsync(String message, Func`1 action) in C:\projects\flow-launcher\Flow.Launcher.Infrastructure\Stopwatch.cs:line 40
   at Flow.Launcher.Core.Plugin.PluginManager.QueryForPluginAsync(PluginPair pair, Query query, CancellationToken token) in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 213
   --- End of inner exception stack trace ---
   at Flow.Launcher.Core.Plugin.PluginManager.QueryForPluginAsync(PluginPair pair, Query query, CancellationToken token) in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 213
   at Flow.Launcher.ViewModel.MainViewModel.<>c__DisplayClass135_0.<<QueryResults>g__QueryTask|2>d.MoveNext() in C:\projects\flow-launcher\Flow.Launcher\ViewModel\MainViewModel.cs:line 821
--- End of stack trace from previous location ---
   at Flow.Launcher.ViewModel.MainViewModel.QueryResults() in C:\projects\flow-launcher\Flow.Launcher\ViewModel\MainViewModel.cs:line 822
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)

```

**Comment by VictoriousRaptor** [comment]
> hmm do you experiment this error?
> 
> ```
> 
> Please open new issue in: https://github.com/Flow-Launcher/Flow.Launcher/issues/new
> 1. upload log file: C:\Users\1\AppData\Roaming\FlowLauncher\Logs\1.11.0\2023-01-22.txt
> 2. copy below exception message
> 
> Flow Launcher version: 1.11.0
> OS Version: Microsoft Windows NT 10.0.22623.0
> IntPtr Length: 8
> x64: True
> 
> Python Path: C:\Users\1\AppData\Roaming\FlowLauncher\Environments\Python\PythonEmbeddable-v3.8.9\pythonw.exe
> Node Path: 
> Date: 01/22/2023 20:16:29
> Exception:
> Explorer Exception: 
> Websites: https://github.com/Flow-Launcher/Flow.Launcher
> Author: Jeremy Wu
> Version: 2.1.0
> Flow.Launcher.Core.ExternalPlugins.FlowPluginException: Object reference not set to an instance of an object.
>  ---> System.NullReferenceException: Object reference not set to an instance of an object.
>    at Flow.Launcher.Plugin.Explorer.Search.SearchManager.<>c__DisplayClass4_0.<SearchAsync>b__1(AccessLink excludedPath) in C:\projects\flow-launcher\Plugins\Flow.Launcher.Plugin.Explorer\Search\SearchManager.cs:line 120
>    at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
>    at System.Collections.Generic.HashSet`1.RemoveWhere(Predicate`1 match)
>    at Flow.Launcher.Plugin.Explorer.Search.SearchManager.SearchAsync(Query query, CancellationToken token) in C:\projects\flow-launcher\Plugins\Flow.Launcher.Plugin.Explorer\Search\SearchManager.cs:line 119
>    at Flow.Launcher.Plugin.Explorer.Main.QueryAsync(Query query, CancellationToken token) in C:\projects\flow-launcher\Plugins\Flow.Launcher.Plugin.Explorer\Main.cs:line 88
>    at Flow.Launcher.Core.Plugin.PluginManager.<>c__DisplayClass22_0.<<QueryForPluginAsync>b__0>d.MoveNext() in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 191
> --- End of stack trace from previous location ---
>    at Flow.Launcher.Infrastructure.Stopwatch.DebugAsync(String message, Func`1 action) in C:\projects\flow-launcher\Flow.Launcher.Infrastructure\Stopwatch.cs:line 40
>    at Flow.Launcher.Core.Plugin.PluginManager.QueryForPluginAsync(PluginPair pair, Query query, CancellationToken token) in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 213
>    --- End of inner exception stack trace ---
>    at Flow.Launcher.Core.Plugin.PluginManager.QueryForPluginAsync(PluginPair pair, Query query, CancellationToken token) in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 213
>    at Flow.Launcher.ViewModel.MainViewModel.<>c__DisplayClass135_0.<<QueryResults>g__QueryTask|2>d.MoveNext() in C:\projects\flow-launcher\Flow.Launcher\ViewModel\MainViewModel.cs:line 821
> --- End of stack trace from previous location ---
>    at Flow.Launcher.ViewModel.MainViewModel.QueryResults() in C:\projects\flow-launcher\Flow.Launcher\ViewModel\MainViewModel.cs:line 822
>    at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
>    at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
>    at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
> ```

looks like similar to #1802
